File: c++/valid-palindrome.cpp
//https://oj.leetcode.com/problems/valid-palindrome/

class Solution {
public:
    bool isPalindrome(string s) {
        const int n = s.size();
        if (n <= 1) return true;
        
        int i = 0, j = n - 1;
        while (i < j) {
            while (i < j && !isAlphaNum(s[i])) i++;
            while (i < j && !isAlphaNum(s[j])) j--;
            if (toLower(s[i]) != toLower(s[j])) return false;
            i++;
            j--;
        }
        
        return true;
    }
    
    bool isAlphaNum(char c) {
        return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
    }
    
    char toLower(char c) {
        if (c >= 'A' && c <= 'Z') {
            c = c - 'A' + 'a';
        }
        return c;
    }
};

File: python/valid-palindrome.py
class Solution:
    # @param s, a string
    # @return a boolean
    def isPalindrome(self, s):
        i = 0
        j = len(s) - 1
        s = s.lower()

        while i < j:
            while i < j and not self._isAlphaNum(s[i]):
                i += 1
            while i < j and not self._isAlphaNum(s[j]):
                j -= 1

            if s[i] != s[j]:
                return False

            i += 1
            j -= 1

        return True


    def _isAlphaNum(self, c):
        return (c >= 'a' and c <= 'z') or (c >= '0' and c <= '9')
